File: list/List.scala
sealed trait List[+A] {

  final def apply(n: Int): Option[A] =
    foldLeft((None: Option[A], n)) {
      case ((None, n), value) if n <= 0 => (Some(value), n)
      case ((result, n), _) => (result, n - 1)
    }._1

  final def append[B >: A](list: List[B]): List[B] =
    this match {
      case Nil => list
      case Cons(head, tail) => Cons(head, tail.append(list))
    }

  final def foldLeft[B](b: B)(f: (B, A) => B): B =
    this match {
      case Cons(head, tail) => tail.foldLeft(f(b, head))(f)
      case Nil => b
    }

  final def size: Int = foldLeft(0)((n, _) => n + 1)

  final def reverse: List[A] = foldLeft(Nil: List[A])((tail, head) => Cons(head, tail))

}

object List {

  def apply[A](values: A*): List[A] =
    values.foldRight(Nil: List[A])(Cons(_, _))

}

case class Cons[A](head: A, tail: List[A]) extends List[A]

case object Nil extends List[Nothing]

object Main extends App {
  val foo: List[Int] = List(0, 1, 2)
  assert(foo.size == 3)
  assert(foo(1) == Some(1))
  assert(foo(2) == Some(2))
  assert(foo(3) == None)
  assert(foo.append(foo) == List(0, 1, 2, 0, 1, 2))
  assert(foo.reverse == List(2, 1, 0))
}

File: list/List.hs
import Control.Exception (assert)

data List a = Nil | Cons a (List a) deriving Eq

foldLeft :: b -> (b -> a -> b) -> List a -> b
foldLeft b _ Nil = b
foldLeft b f (Cons h t) = foldLeft (f b h) f t

size :: List a -> Int
size = foldLeft 0 (\n -> const (n + 1))

nth :: Int -> List a -> Maybe a
nth n = fst . foldLeft (Nothing, n) f
  where
    f (Nothing, n) value | n <= 0 = (Just value, n - 1)
    f (result, n) _ = (result, n - 1)

reversed :: List a -> List a
reversed = foldLeft Nil (flip Cons)

append :: List a -> List a -> List a
append Nil list = list
append (Cons head tail) list = Cons head (append tail list)

list :: [a] -> List a
list = foldr Cons Nil

main :: IO ()
main = let foo = list [0, 1, 2] in
  assert (size foo == 3) $
  assert (nth 1 foo == Just 1) $
  assert (nth 2 foo == Just 2) $
  assert (nth 3 foo == Nothing) $
  assert (nth 3 foo == Nothing) $
  assert (append foo (list [3, 4, 5]) == list [0, 1, 2, 3, 4, 5]) $
  assert (reversed foo == list [2, 1, 0]) $
  return ()
